api/fluentdoperator/v1alpha1/helper.go

```diff
@@ -0,0 +1,258 @@
+)
+
+// +kubebuilder:object:generate=false
+type Renderer interface {
+	GetNamespace() string
+	GetName() string
+	GetCfgId() string
+	GetWatchedLabels() map[string]string
+	GetWatchedNamespaces() []string
+	GetWatchedContainers() []string
+	GetWatchedHosts() []string
+}
+
+// +kubebuilder:object:generate=false
+// Global pluginstores for the fluentd.
+type PluginResources struct {
+	InputPlugins         []params.PluginStore
+	MainRouterPlugins    params.PluginStore
+	LabelPluginResources []params.PluginStore
+}
+
+// +kubebuilder:object:generate=false
+// All the filter/output selected to this cfg
+type CfgResources struct {
+	FilterPlugins []params.PluginStore
+	OutputPlugins []params.PluginStore
+
+	// the hash codes used to depulicate removel
+	FiltersHashcodes map[string]bool
+	OutputsHashcodes map[string]bool
+}
+
+// NewGlobalPluginResources represents a combined global fluentd resources
+func NewGlobalPluginResources(globalId string) *PluginResources {
+	globalMainRouter := fluentdRouter.NewGlobalRouter(globalId)
+	return &PluginResources{
+		InputPlugins:         make([]params.PluginStore, 0),
+		MainRouterPlugins:    *globalMainRouter,
+		LabelPluginResources: make([]params.PluginStore, 0),
+	}
+}
+
+func NewCfgResources() *CfgResources {
+	return &CfgResources{
+		FilterPlugins: make([]params.PluginStore, 0),
+		OutputPlugins: make([]params.PluginStore, 0),
+
+		FiltersHashcodes: make(map[string]bool),
+		OutputsHashcodes: make(map[string]bool),
+	}
+}
+
+func (pgr *PluginResources) CombineGlobalInputsPlugins(sl plugins.SecretLoader, inputs []input.Input) []string {
+	errs := make([]string, 0)
+	for _, f := range inputs {
+		ps, err := f.Params(sl)
+		if err != nil {
+			errs = append(errs, err.Error())
+			continue
+		}
+		pgr.InputPlugins = append(pgr.InputPlugins, *ps)
+	}
+	return errs
+}
+
+func (pgr *PluginResources) BuildCfgRouter(cfg Renderer) (*fluentdRouter.Route, error) {
+	matches := []*fluentdRouter.RouteMatch{
+		{
+			Labels:         cfg.GetWatchedLabels(),
+			Namespaces:     cfg.GetWatchedNamespaces(),
+			Hosts:          cfg.GetWatchedHosts(),
+			ContainerNames: cfg.GetWatchedContainers(),
+		},
+	}
+
+	cfgRoute, err := fluentdRouter.NewRoute(cfg.GetCfgId(), cfg.GetNamespace(), cfg.GetName(), matches)
+	if err != nil {
+		return nil, err
+	}
+
+	// Each fluentd config has its own route plugin
+	routePluginStore, err := cfgRoute.NewRoutePlugin()
+	if err != nil {
+		return nil, err
+	}
+
+	pgr.MainRouterPlugins.InsertChilds(routePluginStore)
+
+	return cfgRoute, nil
+}
+
+// PatchAndFilterClusterLevelResources will combine and patch all the cluster CRs that the fluentdconfig selected,
+// convert the related filter/output pluginstores to the global pluginresources.
+func (pgr *PluginResources) PatchAndFilterClusterLevelResources(sl plugins.SecretLoader, cfgId string,
+	clusterfilters []ClusterFilter, clusteroutputs []ClusterOutput) (*CfgResources, []string) {
```

Comment:
clusterfilters/clusteroutputs/filtersList/outputsList all should be sorted using the same standard to make sure the config generated are the same for every reconcile.

filtersList/outputsList => filters/outputs ?

Reply:
Yeah.